File: Ahorn/entities/Liquid.jl
module XaphanHelperLiquid

using ..Ahorn, Maple

@mapdef Entity "XaphanHelper/Liquid" Liquid(x::Integer, y::Integer,  width::Integer=8, height::Integer=8, lowPosition::Int=0, liquidType::String="acid", frameDelay::Number=0.15, color::String="", transparency::Number=0.65, foreground::Bool=false, riseDelay::Number=0.00, riseDistance::Number=0, riseSpeed::Number=10, riseShake::Bool=false, riseFlag::String="", riseEndFlag::String="", riseSound::Bool=false, directory::String="objects/XaphanHelper/liquid", surfaceHeight::Number=0, visualOnly::Bool=false, canSwim::Bool=false)

const placements = Ahorn.PlacementDict(
    "Liquid (Xaphan Helper)" => Ahorn.EntityPlacement(
        Liquid,
        "rectangle",
        Dict{String, Any}()
    ),
)

Ahorn.editingOptions(entity::Liquid) = Dict{String, Any}(
    "liquidType" => String["acid", "acid_b", "lava", "quicksand", "water"]
)

Ahorn.minimumSize(entity::Liquid) = 16, 8
Ahorn.resizable(entity::Liquid) = true, true

Ahorn.selection(entity::Liquid) = Ahorn.getEntityRectangle(entity)

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::Liquid, room::Maple.Room)
    x = Int(get(entity.data, "x", 0))
    y = Int(get(entity.data, "y", 0))

    width = Int(get(entity.data, "width", 32))
    height = Int(get(entity.data, "height", 32))

    liquid = get(entity.data, "liquidType", "acid")

    colorCode = get(entity.data, "color", "")
    
    if colorCode == ""
        if liquid == "acid"
            colorCode = "88C098"
        end
        if liquid == "acid_b"
            colorCode = "88C098"
        end
        if liquid == "lava"
            colorCode = "F85818"
        end
        if liquid == "quicksand"
            colorCode = "C8B078"
        end
        if liquid == "water"
            colorCode = "669CEE"
        end
    end

    if colorCode != ""
        colorRed = parse(Int, colorCode[1:2], base = 16)
        colorGreen = parse(Int, colorCode[3:4], base = 16)
        colorBlue = parse(Int, colorCode[5:6], base = 16)
        edgeColor = (colorRed, colorGreen, colorBlue, 255) ./ 255
        centerColor = (colorRed, colorGreen, colorBlue, 102) ./ 255

        Ahorn.drawRectangle(ctx, 0, 0, width, height, centerColor, edgeColor)
    end
end

end

File: Loenn/entities/Liquid.lua
local utils = require("utils")

local Liquid = {}

Liquid.name = "XaphanHelper/Liquid"
Liquid.depth = function(room, entity)
    local depth = 0
    if entity.foreground then
        depth = -19999
    else
        depth = -9999
    end
    return depth
end
Liquid.fieldOrder = {
    "x", "y", "width", "height", "liquidType", "directory", "surfaceHeight", "lowPosition", "color", "transparency", "frameDelay", "riseDistance", "riseDelay", "riseSpeed", "riseFlag", "riseEndFlag", "removeFlags", "riseShake", "riseSound", "canSwim", "visualOnly", "foreground"
}
Liquid.fieldInformation = {
    lowPosition = {
        fieldType = "integer",
    },
    liquidType = {
        options = {"acid", "acid_b", "lava", "quicksand", "water"},
        editable = false
    },
    color = {
        fieldType = "color"
    },
    riseDistance = {
        fieldType = "integer",
    },
    riseSpeed = {
        fieldType = "integer",
    },
    surfaceHeight = {
        fieldType = "integer",
        minimumValue = 0
        }
}
Liquid.placements = {
    name = "Liquid",
    data = {
        width = 8,
        height = 8,
        lowPosition = 0,
        liquidType = "acid",
        frameDelay = 0.15,
        color = "88C098",
        transparency = 0.65,
        foreground = false,
        riseDelay = 0.00,
        riseDistance = 0,
        riseSpeed = 10,
        riseShake = false,
        riseFlag = "",
        riseEndFlag = "",
        riseSound = false,
        directory = "objects/XaphanHelper/liquid",
        surfaceHeight = 0,
        canSwim = false,
        visualOnly = false,
        removeFlags = ""
    }
}

local function getEntityColor(entity)
    local defaults = {
        acid = "88C098",
        acid_b = "88C098",
        lava = "F85818",
        quicksand = "C8B078",
        water = "669CEE"
    }
    
    local rawColor = nil
    if not entity.color or entity.color == "" then
        rawColor = defaults[entity.liquidType or "acid"] or "FFFFFF"
    else
        rawColor = entity.color or "FFFFFF"
    end

    local color = utils.getColor(rawColor)

    return color
end

Liquid.fillColor = function(room, entity)
    local color = getEntityColor(entity)

    return {color[1] * 0.3, color[2] * 0.3, color[3] * 0.3, 0.6}
end

Liquid.borderColor = function(room, entity)
    local color = getEntityColor(entity)

    return {color[1] * 0.8, color[2] * 0.8, color[3] * 0.8, 0.8}
end

return Liquid